### PR TITLE
WDL and python scripts to use the VDS in the VAT

### PR DESCRIPTION
**Pull request by RoriCremer**

### CODE REVIEW

File: scripts/variantstore/wdl/GvsCreateVATAnnotations.wdl [hunk 1]
```diff
@@ -243,7 +244,7 @@ task AnnotateVCF {
     # ------------------------------------------------
     # Runtime settings:
     runtime {
-        docker: "annotation/nirvana:3.14"
+        docker: "annotation/nirvana:3.14" # us.gcr.io/broad-dsde-methods/variantstore:nirvana_2022_10_19
```
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATAnnotations.wdl:247]: should be new Nirvana Docker

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+        
+
+
+        # =======================================
+        echo "Creating custom annotations"
+        mkdir customannotations_dir
+        CUSTOM_ANNOTATIONS_FOLDER="$PWD/customannotations_dir"
+
+        # Add AC/AN/AF as custom annotations
+        ## use --skip-ref once you are on a version of nirvana later than 3.14 (once they have created a docker image for it)
```
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:223]:
```suggestion
```

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+    String annotation_json_name = output_annotated_file_name + ".json.gz"
+    String annotation_json_name_jsi = annotation_json_name + ".jsi"
+    String nirvana_location = "/Nirvana/Nirvana.dll"
+    String custom_creation_location = "/Nirvana/SAUtils.dll"
+    String path = "/Cache/GRCh38/Both"
+    String path_supplementary_annotations = "/SupplementaryAnnotation/GRCh38"
+    String path_reference = "/References/Homo_sapiens.GRCh38.Nirvana.dat"
+
+    command <<<
+        # set -e
```
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:202]:
```suggestion
```

File: scripts/variantstore/wdl/GvsCreateVATAnnotations.wdl (hunk 1 above)
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATAnnotations.wdl:247]:
```suggestion
        docker: "us.gcr.io/broad-dsde-methods/variantstore:nirvana_2022_10_19"
```

File: scripts/variantstore/wdl/extract/create_variant_annotation_table.py [hunk 4]
```diff
@@ -177,7 +177,9 @@ def get_subpopulation_calculations(subpop_annotations):
         subpop_an_val = subpop_annotations.get("_".join(["AN", gvs_subpop]), 0)
         subpop_af_val = subpop_annotations.get("_".join(["AF", gvs_subpop]), None)
         # note the assumption is made that AC_Hom must be even because by it's nature it means there are two, but there could be an error
-        subpop_sc_val = int(subpop_annotations.get("_".join(["AC_Hom", gvs_subpop]), 0) / 2 ) + subpop_annotations.get("_".join(["AC_Het", gvs_subpop]), 0) + subpop_annotations.get("_".join(["AC_Hemi", gvs_subpop]), 0)
+        # subpop_sc_val = int(subpop_annotations.get("_".join(["AC_Hom", gvs_subpop]), 0) / 2 ) + subpop_annotations.get("_".join(["AC_Het", gvs_subpop]), 0) + subpop_annotations.get("_".join(["AC_Hemi", gvs_subpop]), 0)
```
Comment (mcovarr) [inline scripts/variantstore/wdl/extract/create_variant_annotation_table.py:180]:
```suggestion
```

File: scripts/variantstore/wdl/extract/create_variant_annotation_table.py [hunk 5]
```diff
@@ -273,7 +275,8 @@ def make_annotated_json_row(row_position, row_ref, row_alt, variant_line, transc
         nirvana_gvs_alleles_fieldname = vat_nirvana_gvs_alleles_dictionary.get(vat_gvs_alleles_fieldname)
         gvs_alleles_fieldvalue = gvs_annotations.get(nirvana_gvs_alleles_fieldname)
         row[vat_gvs_alleles_fieldname] = gvs_alleles_fieldvalue
-        row["gvs_all_sc"] = int(gvs_annotations.get("AC_Hom") / 2 ) + gvs_annotations.get('AC_Het')
+        # row["gvs_all_sc"] = int(gvs_annotations.get("AC_Hom") / 2 ) + gvs_annotations.get('AC_Het') ## should this include the hemi value?
```
Comment (mcovarr) [inline scripts/variantstore/wdl/extract/create_variant_annotation_table.py:278]:
```suggestion
```

File: scripts/variantstore/wdl/extract/extract_subpop.py [hunk 6]
```diff
@@ -52,6 +52,34 @@ def write_custom_annotations_files(observed_subpopulations, custom_annotations_t
         csvout.writerow(description_line)
         csvout.writerow(type_line)
 
+# The above code works for the VDS flavored GVS output
```
Comment (mcovarr) [inline scripts/variantstore/wdl/extract/extract_subpop.py:55]: same issues here with removing + tagging vs commenting out

File: scripts/variantstore/wdl/extract/hail_create_vat_inputs.py [hunk 7]
```diff
@@ -47,54 +65,87 @@ def gq0_to_no_call(vds):
...
+    qc_data = mt.rows() ### TODO: make sure this join does not hard filter!!!!)
+    vd = vds.variant_data
+    vd = vd.annotate_rows(ac_an_af=qc_data[vd.row_key])
+    return hl.vds.VariantDataset(vds.reference_data, vd) ## TODO: IT LOOKS LIKE IT DOES HARD FILTER!
 
 
 def write_sites_only_vcf(vds, sites_only_vcf_path):
     # TODO we will want to drop some cols because there's no reason to carry around some of this stuff
+    split_filtered_vds = hl.vds.split_multi(vds)
+    hl.export_vcf(split_filtered_vds.variant_data.rows(), sites_only_vcf_path)
```
Comment (mcovarr) [inline scripts/variantstore/wdl/extract/hail_create_vat_inputs.py:101]: both lines writing to same place

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl [hunk 8]
```diff
@@ -0,0 +1,739 @@
...
+        DATE=86400 ## 24 hours in seconds
+
+        set +e
+        bq show --project_id ~{project_id} ~{dataset_name}.~{variant_transcript_table} > /dev/null
+        BQ_SHOW_RC=$?
+        set -e
+
+        if [ $BQ_SHOW_RC -ne 0 ]; then
+            echo "Creating a pre-vat table ~{dataset_name}.~{variant_transcript_table}"
+            bq --location=US mk --expiration=$DATE --project_id=~{project_id}  ~{dataset_name}.~{variant_transcript_table} ~{vt_schema}
```
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:356]: makes Aaron sad 😢

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl [hunk 9]
```diff
@@ -0,0 +1,739 @@
...
+            v.clinvar_phenotype,
+        FROM `~{dataset_name}.~{variant_transcript_table}` as v
+            left join
+        (SELECT gene_symbol, ANY_VALUE(gene_omim_id) AS gene_omim_id, ANY_VALUE(omim_phenotypes_id) AS omim_phenotypes_id, ANY_VALUE(omim_phenotypes_name) AS omim_phenotypes_name FROM `~{dataset_name}.~{genes_table}` group by gene_symbol) as g
+        on v.gene_symbol = g.gene_symbol'
+    >>>
+    # ------------------------------------------------
+    # Runtime settings:
+    runtime {
+        docker: "openbridge/ob_google-bigquery:latest"
```
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:513]: use latest gcloud-sdk alpine instead

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl [hunk 10]
```diff
@@ -0,0 +1,739 @@
...
+        cat $(echo $files) > vat_complete.tsv.gz
+        echo_date "bgzipping concatenated file"
+        cat vat_complete.tsv.gz | gunzip | bgzip > vat_complete.bgz.tsv.gz
+        echo_date "copying bgzipped file to ~{output_path}"
+        gsutil -m cp vat_complete.bgz.tsv.gz ~{output_path}
+    >>>
+    # ------------------------------------------------
+    # Runtime settings:
+    runtime {
+        docker: "us.gcr.io/broad-gatk/gatk:4.2.6.1"
```
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:727]: not sure why using gatk Docker here?

File: scripts/variantstore/wdl/extract/create_vat_inputs.py [hunk 11]
```diff
@@ -47,7 +47,7 @@ def download_ancestry_file(gcs_ancestry_file):
     blob = bucket.get_blob(blob_name)
     fd, temp_file = tempfile.mkstemp()
     # Close open descriptor, do not remove temporary file.
-    fd.close()
+    # fd.close()
```
Comment (RoriCremer) [inline scripts/variantstore/wdl/extract/create_vat_inputs.py:50]: Miguel, can you help me figure this bit out?

Comment (mcovarr) [reply scripts/variantstore/wdl/extract/create_vat_inputs.py:50]: I would just delete the line if it's giving you problems, I didn't test this at all 🙈

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl [hunk 12]
```diff
@@ -139,11 +137,172 @@ task MakeSubpopulationFilesAndReadSchemaFiles {
...
+        File custom_annotations_file
+    }
+    String annotation_json_name = output_annotated_file_name + ".json.gz"
+    String annotation_json_name_jsi = annotation_json_name + ".jsi"
+    String nirvana_location = "/Nirvana/Nirvana.dll"
+    String custom_creation_location = "/Nirvana/SAUtils.dll"
+    String path = "/Cache/GRCh38/Both"
+    String path_supplementary_annotations = "/SupplementaryAnnotation/GRCh38"
+    String path_reference = "/References/Homo_sapiens.GRCh38.Nirvana.dat"
+
```
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:191]: parameter_meta {
    nirvana_data_tar: {
      localization_optional: true
    }
}

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl [hunk 13]
```diff
@@ -139,11 +137,172 @@ task MakeSubpopulationFilesAndReadSchemaFiles {
...
+        # Prepend date, time and pwd to xtrace log entries.
+        PS4='\D{+%F %T} \w $ '
+        set -o errexit -o nounset -o pipefail -o xtrace
+
+        # =======================================
+        # Handle our data sources:
+
+        echo "Extracting annotation data sources tar/gzip file..."
+        mkdir datasources_dir
+        tar zxvf ~{nirvana_data_tar} -C datasources_dir  ## --strip-components 2
```
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:204]:
```suggestion
        tar xfz *.tar.gz -C datasources_dir
```

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl [hunk 14]
```diff
@@ -139,11 +137,172 @@ task MakeSubpopulationFilesAndReadSchemaFiles {
...
+    command <<<
+        # set -e
+
+        # Prepend date, time and pwd to xtrace log entries.
+        PS4='\D{+%F %T} \w $ '
+        set -o errexit -o nounset -o pipefail -o xtrace
+
+        # =======================================
+        # Handle our data sources:
+
```
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:201]: gcloud storage cp ~{nirvana_data_tar} .

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl [hunk 15]
```diff
@@ -139,11 +137,172 @@ task MakeSubpopulationFilesAndReadSchemaFiles {
...
+    >>>
+    # ------------------------------------------------
+    # Runtime settings:
+    runtime {
+        docker: "us.gcr.io/broad-dsde-methods/variantstore:nirvana_2022_10_19"
+        memory: "32 GB"
+        cpu: "2"
+        preemptible: 5
+        disks: "local-disk 250 HDD"
```
Comment (mcovarr) [inline scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:240]:
```suggestion
        disks: "local-disk 2000 HDD"
```

File: scripts/variantstore/wdl/extract/create_vat_inputs.py [hunk 16]
```diff
@@ -14,7 +14,7 @@ def parse_ancestry_file(ancestry_file):
     {"ERS4367795":"eur","ERS4367796":"eas","ERS4367797":"eur","ERS4367798":"afr","ERS4367799":"oth", ... }
 
     """
-    start = datetime.now()
+    start = datetime.now() # TODO I don think this is where we want this? The goal here was to see how long the whole VAT pipeline would take
```
Comment (RoriCremer) [inline scripts/variantstore/wdl/extract/create_vat_inputs.py:17]: I think we can clean all of this up? if we are running this in a notebook terminal we likely wont see it anyway

File: scripts/variantstore/wdl/extract/strip_custom_annotations_from_sites_only_vcf.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+
+                    if not chrom_line.endswith("INFO"):
+                        print(f"ERROR: Input Sites only VCF {input_sites_only_vcf} has unexpected header")
+                        sys.exit(1)
+
+                    while True:
+                        vcf_line = input_vcf.readline().rstrip()
+                        if (vcf_line == ""):
+                            print("Done")
+                            sys.exit(0)
```
Comment (mcovarr) [inline scripts/variantstore/wdl/extract/strip_custom_annotations_from_sites_only_vcf.py:22]: might want to restructure this to not `exit` to allow for testing

File: scripts/variantstore/wdl/extract/strip_custom_annotations_from_sites_only_vcf.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+
+                        vcf_fields = vcf_line.split("\t")
+                        info_field_map = get_info_fields(vcf_fields[7])
+
+                        annotations = []
+                        annotations.append(vcf_fields[0])
+                        annotations.append(vcf_fields[1])
+                        annotations.append(vcf_fields[3])
```
Comment (mcovarr) [inline scripts/variantstore/wdl/extract/strip_custom_annotations_from_sites_only_vcf.py:30]: 2 missing on purpose?

Comment (gbggrant) [reply scripts/variantstore/wdl/extract/strip_custom_annotations_from_sites_only_vcf.py]: Yes, field 2 is the 'ID' field in the VCF, not present in the annotations file (nor even populated in our sites only VCF)

File: scripts/variantstore/wdl/extract/strip_custom_annotations_from_sites_only_vcf.py [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    """
+    A method to skip over the header in the VCF
+    :param fp: The file's file pointer
+    :return: The nextmost line that is not a header line
+    """
+    vcf_header = ""
+    while (line := fp.readline().rstrip()):
+        vcf_header += line + '\n'
+        if not line.startswith("##"):
+            return vcf_header, line
```
[the remark this answers is not in the document]
Comment (gbggrant) [reply scripts/variantstore/wdl/extract/strip_custom_annotations_from_sites_only_vcf.py:80]: And it turns out I was wrong here and reverted the behavior to that which I had earlier.

File: scripts/variantstore/wdl/extract/strip_custom_annotations_from_sites_only_vcf.py [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+            sys.exit(1)
+        info_dict[key_value[0]] = key_value[1]
+    return info_dict
+
+
+def skip_vcf_header(fp):
+    """
+    A method to skip over the header in the VCF
+    :param fp: The file's file pointer
+    :return: The nextmost line that is not a header line
```
[the remark this answers is not in the document]
Comment (gbggrant) [reply scripts/variantstore/wdl/extract/strip_custom_annotations_from_sites_only_vcf.py]: Updated / resolved

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl [hunk 21]
```diff
@@ -139,11 +141,259 @@ task MakeSubpopulationFilesAndReadSchemaFiles {
...
+
+    command <<<
+        # set -e
+
+        # Prepend date, time and pwd to xtrace log entries.
+        PS4='\D{+%F %T} \w $ '
+        set -o errexit -o nounset -o pipefail -o xtrace
+
+        # for debugging purposes only
+        gsutil cp ~{annotation_json} '~{output_annotations_gcp_path}' #TODO: Can we run this task without cp-ing down this file?
```
[the remark this answers is not in the document]
Comment (gbggrant) [reply scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:377]: Added this.

File: scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl (hunk 14 above)
Comment (RoriCremer) [reply scripts/variantstore/wdl/GvsCreateVATfromVDS.wdl:301]: so we shouldnt be passing it?

File: scripts/variantstore/wdl/extract/hail_create_vat_inputs.py [hunk 22]
```diff
@@ -38,89 +57,113 @@ def failing_gts_to_no_call(vds):
...
+    Join the dense matrix table input (just AC, AN and AF) back to the original VDS.
     """
-    qc_data = mt.rows()
-    filtered_vd = vds.variant_data
-    filtered_vd = filtered_vd.annotate_rows(ac_an_af=qc_data[filtered_vd.row_key])
-    return hl.vds.VariantDataset(vds.reference_data, filtered_vd)
+    qc_data = mt.rows() ### TODO: make sure this join does not hard filter!!!!)
+    vd = vds.variant_data
+    vd = vd.annotate_rows(ac_an_af=qc_data[vd.row_key])
+    return hl.vds.VariantDataset(vds.reference_data, vd) ## TODO: IT LOOKS LIKE IT DOES HARD FILTER!
```
[the remark this answers is not in the document]
Comment (RoriCremer) [reply scripts/variantstore/wdl/extract/hail_create_vat_inputs.py:107]: we dont even use this method any more, so no stress!

File: scripts/variantstore/wdl/extract/hail_create_vat_inputs.py [hunk 23]
```diff
@@ -38,89 +57,113 @@ def failing_gts_to_no_call(vds):
...
 
 
-def write_sites_only_vcf(vds, sites_only_vcf_path):
-    # TODO we will want to drop some cols because there's no reason to carry around some of this stuff
-    hl.export_vcf(vds.variant_data.rows(), sites_only_vcf_path)
+def write_sites_only_vcf(ac_an_af_split, sites_only_vcf_path):
+    # CHROM	POS	REF	ALT	AC	AN	AF	AC_Hom	AC_Het	AC_Hemi	AC_afr	AN_afr	AF_afr	AC_Hom_afr	AC_Het_afr	AC_Hemi_afr	AC_amr	AN_amr	AF_amr	AC_Hom_amr	AC_Het_amr	AC_Hemi_amr	AC_eas	AN_eas	AF_eas	AC_Hom_eas	AC_Het_eas	AC_Hemi_eas	AC_eur	AN_eur	AF_eur	AC_Hom_eur	AC_Het_eur	AC_Hemi_eur	AC_mid	AN_mid	AF_mid	AC_Hom_mid	AC_Het_mid	AC_Hemi_mid	AC_oth	AN_oth	AF_oth	AC_Hom_oth	AC_Het_oth	AC_Hemi_oth	AC_sas	AN_sas	AF_sas	AC_Hom_sas	AC_Het_sas	AC_Hemi_sas
 
+    pop_info_fields = {}
+    for pop in ['afr', 'amr', 'eas', 'eur', 'mid', 'oth', 'sas']: ## TODO double check that this is all of them
```
[the remark this answers is not in the document]
Comment (RoriCremer) [reply scripts/variantstore/wdl/extract/hail_create_vat_inputs.py:114]: later!

[the remark this answers is not in the document]
Comment (RoriCremer) [reply scripts/variantstore/wdl/extract/hail_create_vat_inputs.py:114]: I should be able to get this from the ancestry file in the future

**Review by gbggrant — APPROVED** [review]

**Pull request merged**